design-documents/atomics-ABI.rst

```diff
@@ -0,0 +1,319 @@
+engineers with such issues.
+
+Backwards Compatibility and New Architecture Features
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Put another way, A baseline ABI helps with the decisions of compatibility of new mappings.
+Certain instructions (such as Load/Store-Pair instructions [ARMARM_]) have different
+single-copy atomicity guarantees with respect to different architecture versions. A baseline
+decides which assembly sequences can be composed correctly (at least as far as testing can decide).
+
+
+Compatibility Between Compilers and Runtimes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The above issues also apply when ensuring object files compiled with different compilers can be mixed. 
+For instance LLVM and GCC code should be interoperable. At the time of writing we identified a number of
+places where this does not apply, both when compiling to target the same architecture version, and when mixing
+different (compatible) architecture versions. Further, the above issues are not limited to statically compiled 
+code. We found one instance where proposed mappings implemented in a JiT compiler would not be interoperable 
+with statically compiled code the runtime links against. Even if a JiT compiles under one set of mappings, and 
+is not subject to an ABI, it may still depend on other libraries or components that do have an ABI.
+
+
+Constrain optimisations
+~~~~~~~~~~~~~~~~~~~~~~~
+
+The frequency of this behaviour justifies collecting these cases together to outline why they should not occur. 
+For example, consider the following Concurrent Program::
+
+  // Shared-Memory Locations
+  _Atomic int* x;
+  _Atomic int* y;
+
+  // Memory Order Parameter
+  #define relaxed memory_order_relaxed
+  #define release memory_order_release
+  #define acquire memory_order_acquire
+
+  // Threads of Execution
+  void thread_0 () {
+    atomic_store_explicit(x,1,relaxed);
+    atomic_thread_fence(release);
+    atomic_store_explicit(y,1,relaxed);
+  }
+
+  void thread_1 () {
+    atomic_exchange_explicit(y,2,release);
+    atomic_thread_fence(acquire);
+    int r0 = atomic_load_explicit(x,relaxed);
+  }
+
+
+Under ISO C, the above Concurrent Program finishes execution in one of three
+possible outcomes (a reference for this notation is found here [PAPER_])::
+
+  { thread_1:r0=0; y=1; }
+  { thread_1:r0=1; y=1; }
+  { thread_1:r0=1; y=2; }
+
+In this case the value read by the exchange on ``thread_1`` is not used, and a
+compiler is free to remove references to unused data. It is not legal according
+to this ABI for a compliant implementation to translate the program into
+the following Assembly Sequences::
+
+  thread_0:
+    MOV W9,#1
+    STR W9,[X2]
+    DMB ISH
+    STR W3,[X4]
+
+  thread_1:
+    MOV W9,#2
+    SWP W9, WZR, [X2]
+    DMB ISHLD
+    LDR W3,[X4]
+
+where ``thread_0:X2`` contains the address of ``x``, ``thread_0:X4`` contains
+the address of ``y``, and ``thread_1:X2`` contains the address of ``y``,
```

Comment:
`and` should move to the last clause:

`the address of ``y``, and ``thread_1:X2`` contains the address of ``y``, ``thread_1:X4`` contains` 

-> `the address of ``y``,  ``thread_1:X2`` contains the address of ``y``, and ``thread_1:X4`` contains`